SentinelMosaicTester/sentinel_mosaic_tester.py

```diff
@@ -0,0 +1,596 @@
+
+  if (scl === 3) { // SC_CLOUD_SHADOW
+    return false;
+  } else if (scl === 9) { // SC_CLOUD_HIGH_PROBA
+    return false;
+  } else if (scl === 8) { // SC_CLOUD_MEDIUM_PROBA
+    return false;
+  } else if (scl === 7) { // SC_CLOUD_LOW_PROBA
+    // return false;
+  } else if (scl === 10) { // SC_THIN_CIRRUS
+    return false;
+  } else if (scl === 11) { // SC_SNOW_ICE
+    return false;
+  } else if (scl === 1) { // SC_SATURATED_DEFECTIVE
+    return false;
+  } else if (scl === 2) { // SC_DARK_FEATURE_SHADOW
+    // return false;
+  }
+  return true;
+}
+
+function evaluatePixel(samples, scenes) {
+  var clo_b02 = [];
+  var clo_b03 = [];
+  var clo_b08 = [];
+  var clo_b02_invalid = [];
+  var clo_b03_invalid = [];
+  var clo_b08_invalid = [];
+  var a = 0;
+  var a_invalid = 0;
+
+  for (var i = 0; i < samples.length; i++) {
+    var sample = samples[i];
+    if (sample.B02 > 0 && sample.B03 > 0 && sample.B08 > 0) {
+      var isValid = validate(sample);
+
+      if (isValid) {
+        clo_b02[a] = sample.B02;
+        clo_b03[a] = sample.B03;
+        clo_b08[a] = sample.B08;
+        a = a + 1;
+      } else {
+        clo_b02_invalid[a_invalid] = sample.B02;
+        clo_b03_invalid[a_invalid] = sample.B03;
+        clo_b08_invalid[a_invalid] = sample.B08;
+        a_invalid = a_invalid + 1;
+      }
+    }
+  }
+
+  var gValue;
+  var bValue;
+  var nValue;
+  if (a > 0) {
+    gValue = getValue(clo_b03);
+    bValue = getValue(clo_b02);
+    nValue = getValue(clo_b08);
+  } else if (a_invalid > 0) {
+    gValue = getValue(clo_b03_invalid);
+    bValue = getValue(clo_b02_invalid);
+    nValue = getValue(clo_b08_invalid);
+  } else {
+    gValue = 0;
+    bValue = 0;
+    nValue = 0;
+  }
+  return {
+    default: [nValue, gValue, bValue]
+  };
+}
+"""
+
+def absolute_to_relative_orbit(x, sat):
+    assert sat in ['2A', '2B']
+    if sat == '2A':
+        adj = -140
+    if sat == '2B':
+        adj = -26
+
+    return (x + adj) % 143
+
+
+def get_dates_by_orbit(bbox, start_date, end_date, max_cc, target_orbit, config):
+    assert target_orbit is not None, "relative_orbit must be specified"
+
+    # convert target_orbit to list if just a single orbit
+    if type(target_orbit) is int:
+        target_orbit = [target_orbit]
+
+    # define time window
+    search_time_interval = (f'{start_date}T00:00:00', f'{end_date}T23:59:59')
+
+    # query scenes
+    wfs_iterator = WebFeatureService(
+        bbox,
+        search_time_interval,
+        data_collection=DataCollection.SENTINEL2_L2A,
+        maxcc=max_cc,
+        config=config
+    )
+
+    # filter down to dates from specified orbit(s)
+    dates = []
+    for tile_info in wfs_iterator:
+        # raw product ID
+        id = tile_info['properties']['id']
+
+        # acquisition date
+        date = tile_info['properties']['date']
+
+        # absolute orbit is buried in ID after _A string
+        absolute_orbit = int(re.search("(?<=_A)[\d]+(?=_)", id).group())
```

Comment:
What does a raw product ID look like? I find the parse pattern more readable generally.

```
S2_TILE_ID = 'S2A_MSIL1C_20180428T142041_N0206_R010_T20JKS_20180428T174250
```

Also looks like we're not guranteed an absolute ID?

_"The components of the Instance Identifier vary slightly when compared to the Product Naming Convention and may include:

The Applicability Start time (_Zzzzzz) = _SYYYYMMDDTHHMMSS Appends the Validity Start Time (granule)

or_

The Absolute Orbit Number (_Zzzzzz) = _Affffff where ffffff is the Absolute Orbit Number (tile)

The Detector ID (_Xxxxxx) = Dxx where xx ranges from 01 to 12 (granule)

or

Tile Number (_Xxxxxx) = _Txxxxx which represents the Tile number according to the US-MGRS naming convention (tile)

The Processing Baseline Number (Nxx.yy) identifies the current processing baseline where x and y may be a digit ranging from 0 to 9."

-->https://sentinels.copernicus.eu/web/sentinel/user-guides/sentinel-2-msi/naming-convention#:~:text=The%20top%2Dlevel%20SENTINEL%2D2,separated%20by%20an%20underscore%20(_).&text=The%20Mission%20Identifier%20(MMM)%20denotes,for%20the%20SENTINEL%2D2B%20instrument.

Reply:
The product IDs that we get from Sentinel Hub have the __granule__ naming convention. There are multiple types of naming convention but that one contains the absolute orbit number.

I got confirmation from Sentinel Hub support that this is the format they return
https://forum.sentinel-hub.com/t/cloudless-sentinel-2-image-sourced-from-a-single-orbit/3294/7?u=hrodman

I agree that the parse syntax is easier to interpret, I will see if I can make it work since the regex is fragile.